libraries/AP_HAL/UARTDriver.cpp

```diff
@@ -226,3 +232,128 @@ void AP_HAL::UARTDriver::log_stats(const uint8_t inst, StatsTracker &stats, cons
 }
 #endif // HAL_LOGGING_ENABLED
 #endif // HAL_UART_STATS_ENABLED
+
+
+/*
+ * Support for logging all incoming/outgoing data from a serial port
+ */
+
+#if HAL_UART_DEBUG_LOGGING_ENABLED
+
+/*
+  log some data for debugging
+*/
+void AP_HAL::UARTDriver::log_read_data(const uint8_t *data, uint16_t length)
+{
+    log_data(data, length, false);
+}
+void AP_HAL::UARTDriver::log_written_data(const uint8_t *data, uint16_t length)
+{
+    log_data(data, length, true);
+}
+void AP_HAL::UARTDriver::log_data(const uint8_t *data, uint16_t length, bool is_written_data)
+{
+    // first see if we actually do want to log:
+    if ((_last_options & OPTION_LOGGING) == 0) {
+        return;
+    }
+
+    // find logging instance corresponding to this UART:
+    if (loginfo == nullptr) {
+        loginfo = NEW_NOTHROW LogInfo();
+        if (loginfo == nullptr) {
+            return;
+        }
+        loginfo->instance = instance;
+        loginfo->next = backend_loginfos;
+        backend_loginfos = loginfo;
+    }
+    if (!log_thread_created) {
+        log_thread_created = true;
+        hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_HAL::UARTDriver::logging_start, void), "uart_log", 4096, AP_HAL::Scheduler::PRIORITY_IO, 0);
+    }
+
+    // write packet into buffer:
+    struct PACKED Header {
+        enum class Flag {
+            IS_WRITTEN_DATA = 1,
+        };
+        uint32_t magic = 0x7fe53b04U;
+        uint32_t time_ms;
+        uint16_t length;
+        uint8_t flags;
+    };
+
+    // note that we are assuming we can do this writing atomically.
+    // If there are multiple writers to the serial port that won't be
+    // the case.  We could reserve, write to and commit an iovec
+    // instead of this:
+
+    const auto space_required = sizeof(Header) + length;
+
+    if (loginfo->buf.space() < space_required) {
+        return;
+    }
+
+    Header header;
+    header.time_ms = AP_HAL::millis();
+    header.length = length;
+    if (is_written_data) {
+        header.flags |= uint8_t(Header::Flag::IS_WRITTEN_DATA);
+    }
+
+    loginfo->buf.write((uint8_t*)&header, sizeof(header));
+    loginfo->buf.write(data, length);
+}
+
+// a linked list of backends which are currently logging:
+AP_HAL::UARTDriver::LogInfo *AP_HAL::UARTDriver::backend_loginfos;
+bool AP_HAL::UARTDriver::log_thread_created;
```

Comment:
Needs a mutex to protect these. The changes in the logging loop will be a bit subtle.

Reply:
You don't need a semaphore for one variable, but there's an implied relation with the struct it points to which matters on non-embedded platforms.

I don't see how it would be benign, any thread could have started the log thread and so we'd end up with multiple randomly writing data to different files.

Thanks for adding it.